### PR TITLE
Add option to drop user data from output

### PR DESCRIPTION
**Pull request by woodpeck**
This pull request adds new outputs (e.g. --xml-clean in addition to --xml) which will create files that omit user IDs and user names, but are otherwise identical. The motivation is preparedness for future, more stringent data protection regulation which might necessitate removing this information from public data. 

An alternative to creating non-userdata versions right away is using software that strips user data from generated files. This is a viable, but more resource intensive option as it requires the parsing of already-generated files. 

The test data in this branch depends on the previous pull request for ae3e4b7 (add historic version to database for tests); applying this to a setup without historic test data would necessitate re-generating the "expected result" files here.

### CODE REVIEW

File: include/xml_writer.hpp [hunk 1]
```diff
@@ -16,7 +16,7 @@ class xml_writer : public output_writer {
 
   xml_writer(const std::string &, const boost::program_options::variables_map &, const user_map_t &,
              const boost::posix_time::ptime &max_time,
-             bool has_history = false, bool has_changeset_discussions = false);
+             bool clean = false, bool has_history = false, bool has_changeset_discussions = false);
```
Comment (zerebubuth) [inline include/xml_writer.hpp]: I'd prefer new default arguments to be appended at the end. This reduces the scope for a change of meaning where invocations specify a few (but not all) arguments. E.g: here, `xml_writer("", vm, user_map, max_time, true)` would mean something different now.

Note that this comment is superseded by one below about moving to `enum class`es anyway.

File: README.md [hunk 2]
```diff
@@ -57,6 +57,9 @@ somewhere with plenty of fast disk space. Existing files may interfere
 with the operation of the program, so it's best to run it in its own,
 clean directory.
 
+All files can be created in a default version (includes "uid" and
+"user" fields), and a "clean" version (without these fields).
```
Comment (zerebubuth) [inline README.md]: I think "clean" has unwarranted positive connotations, and I'd prefer it if this were changed to something which describes the effect more accurately. Perhaps "no-user-info" or "anonymous"?

File: src/pbf_writer.cpp [hunk 3]
```diff
@@ -303,19 +303,23 @@ struct pbf_writer::pimpl {
       info->set_visible(t.visible);
     }
     // set the uid and user information, if the user is public
-    std::map<int64_t, int64_t>::const_iterator itr = m_changeset_user_map.find(t.changeset_id);
-    if (itr != m_changeset_user_map.end()) {
-      user_map_t::const_iterator jtr = m_user_map.find(itr->second);
-      if (jtr != m_user_map.end()) {
-        info->set_uid(jtr->first);
-        info->set_user_sid(str_table(jtr->second));
+    if (!m_clean) {
```
Comment (zerebubuth) [inline src/pbf_writer.cpp]: If the effect of the `m_clean` flag is to make everyone anonymous, could the change be limited to:

```c++
if (m_clean == false && itr != m_changeset_user_map.end()) {
```

Or, if we wish to ensure that `m_changeset_user_map.find` doesn't get called for `m_clean == true` (probably a small optimisation), then something like:

```c++
std::map<int64_t, int64_t>::const_iterator itr = m_changeset_user_map.end();
// when m_clean is set, treat all users as if they were anonymous.
if (m_clean == false) {
  itr = m_changeset_user_map.find(t.changeset_id);
}
if (itr != m_changeset_user_map.end()) {
```

?

File: src/planet-dump.cpp [hunk 4]
```diff
@@ -135,27 +145,63 @@ int main(int argc, char *argv[]) {
     std::vector<boost::shared_ptr<output_writer> > writers;
     if (options.count("history-xml")) {
       std::string output_file = options["history-xml"].as<std::string>();
-      writers.push_back(boost::shared_ptr<output_writer>(new xml_writer(output_file, options, display_name_map, max_time, true)));
+      writers.push_back(boost::shared_ptr<output_writer>(new xml_writer(output_file, options, 
+        display_name_map, max_time, /* clean */ false, /* history */ true)));
```
Comment (zerebubuth) [inline src/planet-dump.cpp]: I think the need to start commenting which booleans are which indicates that we should instead be using `enum class` constants for these. The advantage is two-fold; the values are self-describing, and they are type-safe and cannot be mixed up (as mentioned in a previous comment). For example, something like:

```c++
enum class user_info_level {
  FULL,
  ANON
};

enum class historical_versions {
  NONE,
  FULL
};
```

Then this invocation can be written as:

```c++
writers.push_back(boost::shared_ptr<output_writer>(new xml_writer(output_file, options, 
  display_name_map, max_time, user_info_level::FULL, historical_versions::FULL)));
```

What do you think?

**Comment by woodpeck** [comment]
Your requests make sense and I tried to cater to them. I had to introduce a dependency for C++11 ("warning: scoped enums only available with -std=c++11"). I was unable to nicely handle default parameters so got rid of them altogether but I guess they could be brought back if deemed necessary.

I agree that all this is a bit speculative with a view towards a possible future requirement of publishing userdata-free planets; even if such a requirement arose, it is unclear whether it would perhaps better be satisfied by first creating a normal planet file and then scrubbing it. I wouldn't mind if this pull request got rejected and I'd then simply keep my fork around in case it is needed some day.

**Comment by zerebubuth** [comment]
> it is unclear whether it would perhaps better be satisfied by first creating a normal planet file and then scrubbing it

I think it's probably more efficient to generate them all from the same source. `planet-dump-ng` has already parsed all the input, so to dump it and parse it again would introduce at least some delay.

We might want to have such a program anyway, since presumably we'll need to scrub previous planet files and diffs?

> if such a requirement arose

I hope that we're going to get the chance to discuss whether that would be a requirement.

**Comment by Nakaner** [comment]
@zerebubuth wrote:
> We might want to have such a program anyway, since presumably we'll need to scrub previous planet files and diffs?

You could use the Osmium Tool to remove all metadata except version and timestamp: `osmium cat -o cleaned-planet.osm.pbf --output-format pbf,add_metadata=version+timestamp input.osm.pbf` (latest nightly required because the ability to write only some metadata was added about two weeks ago by Jochen and me. Note that this solution does not work in place but if you clean the dumps one after another, additional 40 GB for a temporary file shouldn't be a problem.

**Pull request merged**